tests/testsuite/artifact_dep.rs

```diff
@@ -2447,3 +2447,375 @@ fn with_assumed_host_target_and_optional_build_dep() {
+        .build();
+    p.cargo("build -Z bindeps")
+        .masquerade_as_nightly_cargo(&["bindeps"])
+        .with_stderr(
+            "\
+[COMPILING] c v0.1.0 ([CWD]/c)
+[COMPILING] b v0.1.0 ([CWD]/b)
+[COMPILING] a v0.1.0 ([CWD]/a)
+[COMPILING] bar v0.1.0 ([CWD]/bar)
+[COMPILING] foo v0.1.0 ([CWD])
+[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn same_target_transitive_dependency_deduplication_lib() {
+    // See https://github.com/rust-lang/cargo/issues/10837
+    let target = rustc_host();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            &format!(
+                r#"
+                [package]
+                name = "foo"
+                version = "0.1.0"
+                edition = "2021"
+
+                [dependencies]
+                a = {{ path = "a" }}
+                b = {{ path = "b", features = ["feature"] }}
+                bar = {{ path = "bar", artifact = "bin", lib = true, target = "{target}" }}
+            "#
+            ),
+        )
+        .file("src/lib.rs", "")
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [package]
+                name = "bar"
+                version = "0.1.0"
+                edition = "2021"
+
+                [dependencies]
+                a = { path = "../a", features = ["b"] }
+                b = { path = "../b" }
+            "#,
+        )
+        .file("bar/src/lib.rs", "")
+        .file(
+            "bar/src/main.rs",
+            r#"
+                use b::Trait;
+
+                fn main() {
+                    a::A.b()
+                }
+            "#,
+        )
+        .file(
+            "a/Cargo.toml",
+            r#"
+                [package]
+                name = "a"
+                version = "0.1.0"
+
+                [dependencies]
+                b = { path = "../b", optional = true }
+            "#,
+        )
+        .file(
+            "a/src/lib.rs",
+            r#"
+                pub struct A;
+
+                #[cfg(feature = "b")]
+                impl b::Trait for A {}
+            "#,
+        )
+        .file(
+            "b/Cargo.toml",
+            r#"
+                [package]
+                name = "b"
+                version = "0.1.0"
+
+                [features]
+                feature = []
+            "#,
+        )
+        .file(
+            "b/src/lib.rs",
+            r#"
+                pub trait Trait {
+                    fn b(&self) {}
+                }
+            "#,
+        )
+        .build();
+    p.cargo("build -Z bindeps")
+        .masquerade_as_nightly_cargo(&["bindeps"])
+        .with_stderr(
+            "\
+[COMPILING] b v0.1.0 ([CWD]/b)
+[COMPILING] a v0.1.0 ([CWD]/a)
+[COMPILING] bar v0.1.0 ([CWD]/bar)
+[COMPILING] foo v0.1.0 ([CWD])
+[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn issue_10525() {
```

Comment:
Maybe a descriptive title.

And other tests should be `decouple_from` instead of `deduplication`, no?

```suggestion
fn same_target_transitive_dependency_decouple_from_hostdep() {
    // See https://github.com/rust-lang/cargo/issues/10525
```

Reply:
Done. I simplified this test case quite a bit and it turned out that it has nothing to do with proc macros and seems to be a duplicate of #11463 #10837
I left the test in regardless, since it has a slightly different structure.

Note, that I uncovered a pretty bizarre inconsistency - if `e` crate is renamed to `a`, the issue does not occur in upstream. How is naming related here is beyond me, but it does seem that there are more things broken the resolver

cc @bstrie

Reply:
Part of my rationale in leaving in the proc macros was to exercise different codepaths as compared to the other test. In addition, while this test case is greatly simplified compared to the original reproducer, care was taken to not simplify it *too* much, in order to produce the same error message as the original maximized bug. I don't mind if it gets simplified further, although this runs the risk of reducing the value of this test as a regression test. I leave the judgment here up to @weihanglo as to whether or not this test is valuable to have, and whether or not it should exercise the proc macro machinery.